include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.cpp

```diff
@@ -0,0 +1,262 @@
+
+    // we only need the gradient
+    integrator_m.evaluate(dealii::EvaluationFlags::gradients);
+    integrator_p.evaluate(dealii::EvaluationFlags::gradients);
+
+    // loop over all quadrature points
+    for(unsigned int q = 0; q < integrator_m.n_q_points; ++q)
+    {
+      // calculate velocity gradient for both elements adjacent to the current face
+      tensor velocity_gradient          = integrator_m.get_gradient(q);
+      tensor velocity_gradient_neighbor = integrator_p.get_gradient(q);
+
+      // get the coefficients
+      scalar viscosity          = this->viscous_kernel->get_coefficient_face(face, q);
+      scalar viscosity_neighbor = this->viscous_kernel->get_coefficient_face_neighbor(face, q);
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+      add_generalized_newtonian_viscosity(viscosity_neighbor, velocity_gradient_neighbor);
+
+      // set the coefficients
+      this->viscous_kernel->set_coefficient_face(face, q, viscosity);
+      this->viscous_kernel->set_coefficient_face_neighbor(face, q, viscosity_neighbor);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::boundary_face_loop_set_coefficients(
+  dealii::MatrixFree<dim, Number> const & matrix_free,
+  VectorType &,
+  VectorType const & src,
+  Range const &      face_range) const
+{
+  FaceIntegratorU integrator(matrix_free,
+                             true,
+                             this->dof_index_velocity,
+                             this->viscous_kernel->get_quad_index());
+
+  // loop over all boundary faces
+  for(unsigned int face = face_range.first; face < face_range.second; face++)
+  {
+    integrator.reinit(face);
+    integrator.read_dof_values(src);
+
+    // we only need the gradient
+    integrator.evaluate(dealii::EvaluationFlags::gradients);
+
+    // loop over all quadrature points
+    for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+    {
+      // calculate velocity gradient
+      tensor velocity_gradient = integrator.get_gradient(q);
+
+      // get the coefficients
+      scalar viscosity = this->viscous_kernel->get_coefficient_face(face, q);
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+
+      // set the coefficients
+      this->viscous_kernel->set_coefficient_face(face, q, viscosity);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::add_generalized_newtonian_viscosity(
+  scalar &       viscosity,
+  tensor const & velocity_gradient) const
+{
+  tensor symmetric_velocity_gradient = 0.5 * (velocity_gradient + transpose(velocity_gradient));
+
+  scalar shear_rate =
+    std::sqrt(2.0 * scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient));
+
+  scalar viscosity_factor = compute_viscosity_factor(shear_rate);
+
+  // add (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  viscosity += viscosity_factor * data.viscosity_margin;
```

Comment:
The comment below says that viscosity margin is (eta_0 - eta_oo), so it could make it more intuitive to understand when changing the comment here, or the implementation (just the order of the terms).

Reply:
Agree! Switched the sequence as the form in the comment is the usual one :100: